### PR TITLE
COMMONSUI Setup

### PR DESCRIPTION
**Pull request by gertie-sheshe**
## Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes # (issue)

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation

### CODE REVIEW

**Review by kilemensi — COMMENTED** [review]
👍🏾 @gertie-sheshe ... Good hustle.

QQ and thoughts

1. Do we need `commons-ui` sub-module in `src` directory? I'm not against it but just want to know what is/was your thinking behind putting it there.
2. Wherever we place it (`src` or not), why not keep the names **exactly** the same i.e. alias `"@commons-ui/core": "whatever-path/commons-ui/packages/core/src"`. This way, the code we write in **PromiseTracker** doesn't have to change one bit.
3. Do we have to `build` `commons-ui` before changes are picked up in **PromiseTracker** source code or what does the `commons.js` script do?

**Comment by gertie-sheshe** [comment]
> 👍🏾 @gertie-sheshe ... Good hustle.
> 
> QQ and thoughts
> 
> 1. Do we need `commons-ui` sub-module in `src` directory? I'm not against it but just want to know what is/was your thinking behind putting it there.
> 2. Wherever we place it (`src` or not), why not keep the names **exactly** the same i.e. alias `"@commons-ui/core": "whatever-path/commons-ui/packages/core/src"`. This way, the code we write in **PromiseTracker** doesn't have to change one bit.
> 3. Do we have to `build` `commons-ui` before changes are picked up in **PromiseTracker** source code or what does the `commons.js` script do?

Thanks @kilemensi 
1. The folder can go anywhere, but I figured having it in the src would be better as it is part of the UI components and the project components folder is on the same folder.
2. I made the names different for the sake of distinction when testing out a component locally. I figured one would want to test only parts of commons-ui? Also, not sure if that will cause an issue with deployment - as it only works with a published version of commons-ui
3. The `commons-build` command basically sets up commons-ui to be able to run locally. The challenge I ran into is with installation of the dependencies. Running `npm i` inside commons-ui and `yarn` on the main project led to errors around having multiple React copies etc. I was able to fix it by combining the package.json files. That's what the commons.js file does. Merges the dependencies, installs, then restores to their initial state.

**Comment by kilemensi** [comment]
Cool @gertie-sheshe ... Can we test using the same names first. Yes, even though one will only be testing small part of COMMONS UI but soon or later someone will forget and push code with "local" names.

**Comment by gertie-sheshe** [comment]
> Cool @gertie-sheshe ... Can we test using the same names first. Yes, even though one will only be testing small part of COMMONS UI but soon or later someone will forget and push code with "local" names.

Yeah - that makes sense @kilemensi 

Hmm..I have just tested it out. Vercel breaks when the alias is renamed. It is no longer linked to the published package

File: README.md [hunk 1]
```diff
@@ -19,6 +19,26 @@ Run the following to get started locally:
...
 
+#### Working with CommonsUI Locally
+
+[CommonsUI](https://github.com/CodeForAfrica/COMMONS-UI) is a library developed and maintained by CfA, that packages React components used to develop various UIs.
+
+To install the commons-ui submodule locally, run:
+
+```sh
+  yarn submodules
+  npm run commons-setup
```
Comment (kilemensi) [inline README.md:30]: If we're using `yarn`, lets use `yarn` everywhere ... no need to mix it up with `npm`

File: .gitmodules [hunk 2]
```diff
@@ -0,0 +1,3 @@
+[submodule "commons-ui"]
+	path = commons-ui
```
Comment (kilemensi) [inline .gitmodules:2]:
```suggestion
	path = COMMONS-UI
```

File: README.md [hunk 3]
```diff
@@ -19,6 +19,26 @@ Run the following to get started locally:
...
+```
+
+CommonsUI is now ready to be tested locally. To get started, check out to the branch you want to test:
```
Comment (kilemensi) [inline README.md:33]:
```suggestion
COMMONS-UI is now ready to be tested locally. To get started, check out to the branch you want to test:
```

File: commons.js [hunk 4]
```diff
@@ -0,0 +1,19 @@
+const fs = require("fs");
+
+const packageOne = require("./package.json");
+const packageTwo = require("./commons-ui/package.json");
```
Comment (kilemensi) [inline commons.js:4]:
```suggestion
const mainPackage = require("./package.json");
const commonsUIPackage = require("./commons-ui/package.json");
```

File: commons.js [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+const babel = require("./.babelrc.json");
+
+// Merge dependencies for install
+Object.assign(packageOne.devDependencies, packageTwo.devDependencies);
+Object.assign(packageOne.devDependencies, packageOne.dependencies);
```
Comment (kilemensi) [inline commons.js:9]: Can you may be add a comment why we need this override here?

**Review by kilemensi — COMMENTED** [review]
👍🏾 @gertie-sheshe ... In addition to the changes we discussed:

All scripts should be in a `scripts` or `script` folder ... giving them _unix-y_ type of names e.g. _commons-ui-add_, _commons-ui-remove_, etc.

File: README.md [hunk 6]
```diff
@@ -19,6 +19,29 @@ Run the following to get started locally:
 
 **Note:** Reach out to your team lead for directions on how to receive the `CHECK_ACCESS_TOKEN`
 
+#### Working with CommonsUI Locally
+
+[CommonsUI](https://github.com/CodeForAfrica/COMMONS-UI) is a library developed and maintained by CfA, that packages React components used to develop various UIs.
```
Comment (kilemensi) [inline README.md:24]:
```suggestion
[COMMONS UI](https://github.com/CodeForAfrica/COMMONS-UI) is a library, developed and maintained by CfA, that packages React components used to develop various UIs.
```

File: README.md [hunk 7]
```diff
@@ -19,6 +19,29 @@ Run the following to get started locally:
...
+To build and run the project with COMMONS-UI submodule locally, run:
+
+```sh
+  yarn dev:commons
+```
+
+COMMONS-UI is now ready to be tested locally. To get started, check out to the branch you want to test:
+
+```sh
+  cd src/commons-ui
```
Comment (kilemensi) [inline README.md:35]: Is this still the case?

File: package.json [hunk 8]
```diff
@@ -62,22 +62,26 @@
...
     "dev:next": "next dev",
+    "submodule": "git submodule init && git submodule update",
     "export": "yarn run build && next export",
     "format": "yarn prettier --write --ignore-path './.gitignore' \"**/*.+(json|md|yml)\" './'",
     "format-staged": "prettier --write",
     "lint": "yarn eslint --fix --ignore-path './.gitignore' './'",
     "lint-staged": "yarn eslint --fix --ext '.js,.json,.md,.yml'",
-    "start": "next start"
+    "start": "next start",
+    "commons-setup": "node ./scripts/commons-ui-add.js && yarn && git restore package.json && git restore yarn.lock",
```
Comment (kilemensi) [inline package.json:75]:
```suggestion
    "setup:commons": "node ./scripts/commons-ui-add.js && yarn && git restore package.json && git restore yarn.lock",
```
Or `"add:commons"`, etc., i.e. it should be a verb.

**Review by kilemensi — APPROVED** [review]
🚀 @gertie-sheshe ... looks ready to me.

**Pull request merged**